File: meta_ads/audiences.py
"""Custom Audience durum kontrolu.

Bir kitleyi kampanyada hedeflemeden once "kullanima hazir mi" sorusunu
yanitlar - Meta yeni olusturulan/genisleyen kitleleri belli bir boyuta
ulasana kadar "too small" olarak isaretler ve kampanyada kullanmaya izin
vermez.
"""
from __future__ import annotations

from facebook_business.adobjects.customaudience import CustomAudience

from .client import call, ensure_api_initialized

# delivery_status.code == 200 -> aktif/hazir, 300 -> kucuk/pasif.
# operation_status.code == 200 -> normal islem durumu.
DELIVERY_READY_CODE = 200
DELIVERY_TOO_SMALL_CODE = 300


def get_status(audience_id: str) -> dict:
    """Kitlenin guncel boyut/durumunu Meta'dan ceker."""
    ensure_api_initialized()
    audience = CustomAudience(audience_id)
    fields = [
        CustomAudience.Field.name,
        CustomAudience.Field.approximate_count_lower_bound,
        CustomAudience.Field.approximate_count_upper_bound,
        CustomAudience.Field.operation_status,
        CustomAudience.Field.delivery_status,
    ]
    data = call(audience.api_get, fields=fields)

    delivery = data.get(CustomAudience.Field.delivery_status) or {}
    delivery_code = delivery.get("code")
    lower = data.get(CustomAudience.Field.approximate_count_lower_bound)

    ready = delivery_code == DELIVERY_READY_CODE

    return {
        "id": audience_id,
        "name": data.get(CustomAudience.Field.name),
        "approx_size_lower": lower,
        "approx_size_upper": data.get(CustomAudience.Field.approximate_count_upper_bound),
        "operation_status": data.get(CustomAudience.Field.operation_status),
        "delivery_status": delivery,
        "ready": ready,
    }


def ensure_ready(audience_id: str) -> dict:
    """Kitle hazir degilse ValueError firlatir - kampanya olusturmadan once cagirilir."""
    status = get_status(audience_id)
    if not status["ready"]:
        delivery = status["delivery_status"] or {}
        reason = delivery.get("description", "durum bilinmiyor")
        raise ValueError(
            f"Kitle '{status['name']}' ({audience_id}) henuz kampanyada kullanima hazir degil.\n"
            f"Meta'nin sebebi: {reason} (yaklasik boyut: {status['approx_size_lower']}).\n"
            "Kitle dolana kadar bekle (genelde birkac saat/gun) ya da 'uygulamayi acan herkes 180' "
            "gibi zaten hazir baska bir kitle kullan."
        )
    return status

File: worker/src/meta/audiences.ts
/**
 * Custom Audience durum kontrolu - meta_ads/audiences.py'nin portu.
 *
 * Bir kitleyi kampanyada hedeflemeden once "kullanima hazir mi" sorusunu
 * yanitlar - Meta yeni olusturulan/genisleyen kitleleri belli bir boyuta
 * ulasana kadar "too small" olarak isaretler ve kampanyada kullanmaya izin
 * vermez.
 */
import { adAccountPath, graphRequest, graphRequestPaged, type MetaEnv } from "./client";

const DELIVERY_READY_CODE = 200;

export interface AudienceStatus {
  id: string;
  name: string | null;
  approxSizeLower: number | null;
  approxSizeUpper: number | null;
  operationStatus: unknown;
  deliveryStatus: { code?: number; description?: string } | null;
  ready: boolean;
}

export async function getAudienceStatus(env: MetaEnv, audienceId: string): Promise<AudienceStatus> {
  const fields = [
    "name",
    "approximate_count_lower_bound",
    "approximate_count_upper_bound",
    "operation_status",
    "delivery_status",
  ].join(",");
  const data = await graphRequest(env, `/${audienceId}`, { params: { fields } });
  const delivery = data.delivery_status ?? null;

  return {
    id: audienceId,
    name: data.name ?? null,
    approxSizeLower: data.approximate_count_lower_bound ?? null,
    approxSizeUpper: data.approximate_count_upper_bound ?? null,
    operationStatus: data.operation_status ?? null,
    deliveryStatus: delivery,
    ready: delivery?.code === DELIVERY_READY_CODE,
  };
}

export async function listAudiences(
  env: MetaEnv,
  opts: { limit?: number } = {},
): Promise<Record<string, unknown>[]> {
  const fields = [
    "id",
    "name",
    "approximate_count_lower_bound",
    "approximate_count_upper_bound",
    "delivery_status",
  ].join(",");
  return await graphRequestPaged(env, adAccountPath(env, "customaudiences"), {
    fields,
    limit: opts.limit ?? 50,
  });
}

/** Kitle hazir degilse hata firlatir - kampanya olusturmadan once cagirilir. */
export async function ensureAudienceReady(env: MetaEnv, audienceId: string): Promise<AudienceStatus> {
  const status = await getAudienceStatus(env, audienceId);
  if (!status.ready) {
    const reason = status.deliveryStatus?.description ?? "durum bilinmiyor";
    throw new Error(
      `Kitle '${status.name}' (${audienceId}) henuz kampanyada kullanima hazir degil.\n` +
        `Meta'nin sebebi: ${reason} (yaklasik boyut: ${status.approxSizeLower}).\n` +
        "Kitle dolana kadar bekle (genelde birkac saat/gun) ya da zaten hazir baska bir kitle kullan.",
    );
  }
  return status;
}
